mitxgraders/helpers/specify_domain.py

```diff
@@ -0,0 +1,241 @@
+    DomainError: There was an error evaluating function cross(...)
+    ...1st input is ok: received a vector of length 3 as expected
+    ...2nd input has an error: received a vector of length 2, expected a vector of length 3
+    >>> cross(a)
+    Traceback (most recent call last):
+    DomainError: There was an error evaluating function cross(...): expected 2 inputs, but received 1.
+    """
+
+    schema_config = Schema({
+        Required('input_shapes'): [SCHEMA_SHAPE],
+        Required('display_name', default=None): str
+    })
+
+    def __init__(self, config=None, **kwargs):
+        super(SpecifyDomain, self).__init__(config, **kwargs)
+
+        display_name = self.config['display_name']
+        shapes = self.config['input_shapes']
+        self.decorate = self.make_decorator(*shapes, display_name=display_name)
+
+    def __call__(self, func):
+        return self.decorate(func)
+
+    @staticmethod
+    def make_decorator(*shapes, **kwargs):
+        """
+        Constructs the decorator that validates inputs.
+
+        This method is NOT author-facing; its inputs undero no validation.
+
+        Used internally in mitxgraders library.
+        """
+
+        display_name = kwargs.get('display_name', None)
+        schemas = [Schema(has_shape(shape)) for shape in shapes]
+
+        # can't use @wraps, func might be a numpy ufunc
+        def decorator(func):
+            func_name = display_name if display_name else func.__name__
+            def _func(*args):
+                if len(shapes) != len(args):
+                    msg = ("There was an error evaluating function "
+                           "{func_name}(...): expected {expected} inputs, but "
+                           "received {received}."
+                           .format(func_name=func_name, expected=len(shapes), received=len(args))
+                          )
+                    raise DomainError(msg)
+
+                errors = []
+                for schema, arg in zip(schemas, args):
+                    try:
+                        schema(arg)
+                        errors.append(None)
+                    except Invalid as error:
+                        errors.append(error)
+
+                if all([error is None for error in errors]):
+                    return func(*args)
+
+                lines = ['There was an error evaluating function {0}(...)'.format(func_name)]
+                for index, (shape, error) in enumerate(zip(shapes, errors)):
+                    ordinal = low_ordinal(index + 1)
+                    if error:
+                        lines.append('{0} input has an error: '.format(ordinal) + error.error_message)
+                    else:
+                        expected = 'scalar' if shape == (1,) else MathArray.get_description(shape)
+                        lines.append('{0} input is ok: received a {1} as expected'
+                            .format(ordinal, expected))
+
+                message = "\n\t".join(lines)
```

Comment:
Tab characters look fine in bash, but they don't render in HTML. Do you plan to modify log output to transform tab characters into something meaningful?

Reply:
Good point. I just removed them, it made the doc tests fragile anyway.